ui/src/components/Canvas.tsx

```diff
@@ -1090,17 +1123,150 @@ export function Canvas() {
   const [client, setClient] = useState({ x: 0, y: 0 });
 
   const onPaneContextMenu = (event) => {
+    console.log("onPaneContextMenu", event);
     event.preventDefault();
     setShowContextMenu(true);
     setPoints({ x: event.pageX, y: event.pageY });
     setClient({ x: event.clientX, y: event.clientY });
+    console.log(showContextMenu, points, client);
   };
 
+  const pasteCodePod = useCallback(
+    ({ x, y, pod }) => {
+      const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();
+      let [posX, posY] = [
+        reactFlowBounds.width / 2,
+        reactFlowBounds.height / 2,
+      ];
+      const position = reactFlowInstance.project({ x: posX, y: posY });
+      position.x = (position.x - pod.width! / 2) as number;
+      position.y = (position.y - (pod.height ?? 0) / 2) as number;
+      console.log("pasteCodePod", position, clientId, pod);
+
+      const style = {
+        width: pod.width,
+        height: undefined,
+        opacity: 0.5,
+      };
+
+      const id = nanoid();
+      const newNode = {
+        id,
+        type: "code",
+        position,
+        data: {
+          name: pod?.name || "",
+          label: id,
+          parent: "ROOT",
+          clientId,
+        },
+        level: 0,
+        extent: "parent",
+        parentNode: undefined,
+        dragHandle: ".custom-drag-handle",
+        style,
+      };
+
+      addPod(apolloClient, {
+        id,
+        parent: "ROOT",
+        type: "CODE",
+        lang: "python",
+        x: position.x,
+        y: position.y,
+        width: pod.width,
+        height: pod.height,
+        content: pod.content,
+        error: pod.error,
+        stdout: pod.stdout,
+        result: pod.result,
+        name: pod.name,
+        dirty: true,
+      });
+
+      nodesMap.set(id, newNode as any);
+      setPasting(id);
```

Comment:
If we set nodesMap, it would automatically sync with collaborators, right? UPDATE resolved; I saw that you checked `clientID` in various places in `lib/nodes.tsx`. That's a clever trick, looks good to me!

Reply:
Yes, it is an adding event in `nodesMap`, [`ymap.observe`](https://docs.yjs.dev/api/shared-types/y.map#observing-changes-y.mapevent) triggers a `addPod` calls without remote DB writing, just adds it to the `store.pods` in each user's side. In this case, since I add the checking statement for `clientId`, collaborators will ignore it and won't call `addPod` until we delete the temporary moving node and re-add the formal node.